File: vibeshield-python/examples/command_usage.py
"""
VibeShield Command Injection Protection and safe-exec — Usage Example

Demonstrates how to sanitize command arguments, validate commands, and execute commands safely.

Run with: python examples/command_usage.py
"""

import os
import sys
import platform

# Allow importing from src/
sys.path.insert(0, os.path.join(os.path.dirname(__file__), ".."))

from src import (
    sanitize_shell_input,
    validate_safe_command,
    safe_exec,
    VibeShieldCommandInjectionError
)

print("=============================================================")
print("[SECURE] VIBESHIELD COMMAND INJECTION PROTECTION DEMO")
print("=============================================================\n")

# 1. Sanitizing inputs
print("--- 1. Using sanitize_shell_input() ---")
test_inputs = [
    "safe-arg123",
    "user@domain.com",
    "malicious_arg; rm -rf /",
    "some_arg\x00_inject",
    "another_arg|ls",
    "backtick`inject`",
]

for inp in test_inputs:
    try:
        sanitized = sanitize_shell_input(inp)
        print(f"[SAFE] Input: '{inp}' -> Sanitized: '{sanitized}'")
    except VibeShieldCommandInjectionError as e:
        print(f"[BLOCKED] Input: '{inp}' -> Error: {e}")
print()

# 2. Validating commands
print("--- 2. Using validate_safe_command() ---")
test_commands = [
    ("ping", ["127.0.0.1"]),
    ("bash", ["-c", "ls"]),
    ("../ping", ["127.0.0.1"]),
    ("ping", ["../etc/passwd"]),
    ("ping", ["C:\\some\\path"]),
]

for cmd, args in test_commands:
    try:
        validate_safe_command(cmd, args)
        print(f"[SAFE] Command: {cmd} with args {args}")
    except VibeShieldCommandInjectionError as e:
        print(f"[BLOCKED] Command: {cmd} with args {args} -> Error: {e}")
print()

# 3. Executing commands safely
print("--- 3. Using safe_exec() ---")
try:
    count_flag = "-n" if platform.system().lower() == "windows" else "-c"
    print("Running a safe ping...")
    result = safe_exec("ping", [count_flag, "1", "127.0.0.1"])
    print(f"Success! Exit Code: {result['code']}")
    # print first 5 lines of stdout
    lines = result['stdout'].splitlines()[:5]
    print("Output preview:")
    for line in lines:
        print(f"  {line}")
except Exception as e:
    print(f"Execution failed: {e}")

print("\nAttempting to execute an unauthorized command...")
try:
    safe_exec("rm", ["-rf", "tmp"])
except VibeShieldCommandInjectionError as e:
    print(f"[BLOCKED] Correctly intercepted: {e}")

File: vibeshield-ts/examples/command-usage.ts
/**
 * VibeShield Command Injection Protection — Usage Example
 *
 * This file demonstrates how to use the Command Injection Protection utilities and
 * the secure safeExec wrapper to safely run external binaries.
 *
 * Run with: npx tsx examples/command-usage.ts
 */

import {
  sanitizeShellInput,
  validateSafeCommand,
  safeExec,
  VibeShieldCommandInjectionError
} from '../src/index.js';

console.log('=============================================================');
console.log('🛡️ VIBESHIELD COMMAND INJECTION PROTECTION DEMO');
console.log('=============================================================\n');

// 1. Sanitize Shell Input
console.log('--- 1. Sanitizing Arguments ---');
const inputs = [
  'google.com',
  'api.v1.host.org',
  '--help',
  '-v',
  'google.com; rm -rf /',
  'google.com | id',
  '$(whoami)',
  'google.com\0',
  'google.com\nwhoami'
];

for (const input of inputs) {
  try {
    const sanitized = sanitizeShellInput(input);
    console.log(`✅ Input: "${input}" -> Sanitized: "${sanitized}"`);
  } catch (error) {
    if (error instanceof VibeShieldCommandInjectionError) {
      console.log(`❌ Input: "${input}" -> Rejected: ${error.message}`);
    } else {
      console.log(`💥 Unknown Error: ${(error as any).message}`);
    }
  }
}
console.log();

// 2. Validate Command & Arguments
console.log('--- 2. Validating Commands & Arguments ---');
const commandScenarios = [
  { command: 'ping', args: ['8.8.8.8'] },
  { command: 'rm', args: ['-rf', '/'] },
  { command: 'ping', args: ['../../etc/passwd'] },
  { command: '/usr/bin/ping', args: ['8.8.8.8'] },
  { command: 'curl', args: ['example.com'] },
  { command: 'git', args: ['status'] }
];

for (const scenario of commandScenarios) {
  try {
    validateSafeCommand(scenario.command, scenario.args);
    console.log(`✅ Safe Command: "${scenario.command} ${scenario.args.join(' ')}"`);
  } catch (error) {
    if (error instanceof VibeShieldCommandInjectionError) {
      console.log(`❌ Unsafe Command: "${scenario.command} ${scenario.args.join(' ')}" -> Rejected: ${error.message}`);
    } else {
      console.log(`💥 Unknown Error: ${(error as any).message}`);
    }
  }
}
console.log();

// 3. Executing commands safely
console.log('--- 3. Executing Commands Safely ---');

async function runDemo() {
  // We can try to execute a safe ping command
  // On Windows, the default allowed ping might run. Let's try to execute a short ping.
  const target = '127.0.0.1';
  console.log(`Executing safe command: ping ${target}`);
  try {
    const result = await safeExec('ping', [target]);
    console.log(`✅ Execution Completed (exit code: ${result.code})`);
    console.log(`stdout:\n${result.stdout}`);
  } catch (error) {
    console.log(`❌ Execution failed: ${(error as any).message}`);
  }

  console.log('\nTrying to execute a blocked command: rm -rf /');
  try {
    await safeExec('rm', ['-rf', '/']);
    console.log('✅ Executed? (Should not happen!)');
  } catch (error) {
    if (error instanceof VibeShieldCommandInjectionError) {
      console.log(`❌ Successfully Blocked: ${error.message}`);
    } else {
      console.log(`💥 Unknown Error: ${(error as any).message}`);
    }
  }

  console.log('\n=============================================================');
  console.log('💡 TIP: Use safeExec() to invoke binaries with shell: false.');
  console.log('   Always filter commands through validateSafeCommand().');
  console.log('=============================================================');
}

runDemo();
